pkg/loki/config_wrapper.go

```diff
@@ -96,30 +99,54 @@ func (c *ConfigWrapper) ApplyDynamicConfig() cfg.Source {
 	}
 }
 
-// applyIngesterRingConfig will use whatever config is setup for the ingester ring and use it everywhere else
-// we have a ring configured. The reason for centralizing on the ingester ring as this is been set in basically
-// all of our provided config files for all of time, usually set to `inmemory` for all the single binary Loki's
-// and is the most central ring config for Loki.
+// applyDynamicRingConfigs checks the current config and, depending on the given values, reuse ring configs accordingly.
 //
-// If the ingester ring has its interface names sets to a value equal to the default (["eth0", en0"]), it will try to append
-// the loopback interface at the end of it.
-func applyIngesterRingConfig(r *ConfigWrapper, defaults *ConfigWrapper) error {
-	if reflect.DeepEqual(r.Ingester.LifecyclerConfig.InfNames, defaults.Ingester.LifecyclerConfig.InfNames) {
-		appendLoopbackInterface(r)
+// 1. Gives preference to any explicit ring config set. For instance, if the user explicitly configures Distributor's ring,
+// that config will prevail.
+// 2. If no explicit ring config is set, reuse the common ring configured if it was provided.
+// 3. If no common ring was provided, reuse whatever was set by the ingester ring.
+func applyDynamicRingConfigs(r, defaults *ConfigWrapper) error {
+	// if common/ring is NOT set, reuse ingester ring config.
+	if reflect.DeepEqual(r.Common.Ring, defaults.Common.Ring) {
+		reuseIngesterConfig(r, defaults)
+		return nil
 	}
 
-	lc := r.Ingester.LifecyclerConfig
-	rc := r.Ingester.LifecyclerConfig.RingConfig
-	s := rc.KVStore.Store
-	sc := r.Ingester.LifecyclerConfig.RingConfig.KVStore.StoreConfig
+	return reuseCommonConfig(r, defaults)
+}
 
-	f, err := tokensFile(r, "ingester.tokens")
+func reuseCommonConfig(r, defaults *ConfigWrapper) error {
```

Comment:
this could use a better name, as it's not really "reusing", and we should call out it applies specifically to the ring